# Tugas Metode Numerik - Penyelesaian SPL dengan Metode Relaksasi

**Kelompok 3 - Kelas D Informatika**

## Anggota Kelompok

-  Muchammad Yuda Tri Ananda (24060124110142)
-  Novelya Cherina (24060124140174) 
-  Nashwa Aldebaran (24060124140150)

# Metode Relaksasi (Successive Over Relaxation - SOR)

Successive Over Relaxation (SOR) merupakan varian dari Metode Gauss-Siedel yang dapat digunakan untuk meningkatkan konvergensi. Cara kerjanya adalah mengambil estimasi $k$ dari Gauss-Siedel untuk memperoleh estimasi $k+1$ menggunakan parameter relaksasi ω (menentukan sejauh mana estimasi baru didasarkan pada estimasi sebelumnya).

𝜔 ∈ (0,2) → faktor relaksasi
- 𝜔 = 1, sama dengan Gauss Seidel
- 𝜔 > 1, over relaksasi (SOR) → konvergensi cepat
- 𝜔 < 1, under relaksasi (SUR) → konvergensi stabil

Dalam metode SOR kita gunakan konvergensi cepat dengan tujuan mendapatkan hasil yang konvergen.

Persamaan SOR sebagai berikut.

$$x_i^{(k+1)}=(1-ω) x_i^{(k)}+\frac{ω}{a_{ii}}(b_i - \sum_{j<i} a_{ij}x_j^{(k+1)} - \sum_{j>i} a_{ij}x_j^{(k)})$$

Atau bisa juga ditulis sebagai:
$$x^{(k+1)}=(1-ω) x^{(k)}+ω∙[Persamaan GS]$$
Dimana $[Persamaan GS]$ adalah ekspresi untuk $x_i$ dari metode Gauss-Seidel.

## Contoh Kasus: Load Balancing Server

Diberikan sebuah kasus Load Balancing pada jaringan komputer. Sebuah sistem cloud terdiri dari dua server utama: Server A dan Server B, yang bekerja secara paralel untuk menangani permintaan pengguna dari dua aplikasi besar, App1 dan App2.

- Untuk App1, beban kerja didominasi oleh Server A, dengan rasio alokasi 10:1 terhadap Server B. Total beban App1 adalah 27 unit.
- Untuk App2, beban kerja didominasi oleh Server B, dengan rasio alokasi 1:10 terhadap Server A. Total beban App2 adalah 29 unit.

Tentukan beban ideal pada setiap Unit!

Misal:
- `x` = beban kerja (unit) yang ditangani Server A
- `y` = beban kerja (unit) yang ditangani Server B

Maka Sistem Persamaan Linier (SPL) nya adalah:
$$10x + y = 27$$
$$x + 10y = 29$$

## Langkah 1 - Ubah dalam bentuk persamaan Gauss-Seidel

Dari SPL:
$$10x + y = 27 \quad (1)$$
$$x + 10y = 29 \quad (2)$$

Kita dapat mengekspresikan x dari persamaan (1) dan y dari persamaan (2) sebagai berikut:
$$x = \frac{27 - y}{10}$$
$$y = \frac{29 - x}{10}$$

Dalam bentuk iteratif Gauss-Seidel, persamaannya menjadi:
$$x^{(k+1)} = \frac{27 - y^{(k)}}{10}$$
$$y^{(k+1)} = \frac{29 - x^{(k+1)}}{10}$$

In [ ]:
# Definisikan fungsi untuk bagian Gauss-Seidel dari persamaan SOR
f1_gs = lambda x,y: (27 - y) / 10
f2_gs = lambda x,y: (29 - x) / 10

## Langkah 2 - Inisialisasi $k=0$

Inisialisasikan nilai awal, dimana:
$$x^{(0)} = 0$$
$$y^{(0)} = 0$$

## Langkah 3 - Susun dalam bentuk persamaan SOR

Dengan persamaan SOR:
$$x_i^{(k+1)}=(1-ω) x_i^{(k)}+ω∙[Persamaan\ GS_i]$$

Kita kemudian memasukkan persamaan Gauss-Seidel yang sudah kita buat sebelumnya, menjadi seperti berikut.
$$x^{(k+1)}=(1-ω) x^{(k)}+ω\left[\frac{27-y^{(k)}}{10}\right]$$
$$y^{(k+1)}=(1-ω) y^{(k)}+ω\left[\frac{29-x^{(k+1)}}{10}\right]$$
(Perhatikan bahwa untuk $y^{(k+1)}$, kita menggunakan nilai $x^{(k+1)}$ yang baru saja dihitung pada iterasi yang sama).

## Langkah 4 - Iterasi (Contoh Manual)

Kita misalkan $ω=1.05$ sehingga bentuk persamaan menjadi seperti berikut.
$$x^{(k+1)}=(1-1.05) x^{(k)}+1.05\left[\frac{27-y^{(k)}}{10}\right] = -0.05 x^{(k)} + 1.05\left[\frac{27-y^{(k)}}{10}\right]$$
$$y^{(k+1)}=(1-1.05) y^{(k)}+1.05\left[\frac{29-x^{(k+1)}}{10}\right] = -0.05 y^{(k)} + 1.05\left[\frac{29-x^{(k+1)}}{10}\right]$$

### Iterasi 1
Dengan $x^{(0)} = 0$ dan $y^{(0)} = 0$, dan $ω=1.05$:

Nilai x:
$$x^{(1)} = -0.05 x^{(0)} + 1.05\left[\frac{27-y^{(0)}}{10}\right]$$
$$x^{(1)} = -0.05(0) + 1.05\left[\frac{27-0}{10}\right] = 1.05 \times 2.7 = 2.835$$

Nilai y (menggunakan $x^{(1)}=2.835$):
$$y^{(1)} = -0.05 y^{(0)} + 1.05\left[\frac{29-x^{(1)}}{10}\right]$$
$$y^{(1)} = -0.05(0) + 1.05\left[\frac{29-2.835}{10}\right] = 1.05\left[\frac{26.165}{10}\right] = 1.05 \times 2.6165 = 2.747325$$

### Iterasi 2
Dengan $x^{(1)} = 2.835$ dan $y^{(1)} = 2.747325$, dan $ω=1.05$:

Nilai x:
$$x^{(2)} = -0.05 x^{(1)} + 1.05\left[\frac{27-y^{(1)}}{10}\right]$$
$$x^{(2)} = -0.05(2.835) + 1.05\left[\frac{27-2.747325}{10}\right] = -0.14175 + 1.05\left[\frac{24.252675}{10}\right]$$
$$x^{(2)} = -0.14175 + 1.05 \times 2.4252675 = -0.14175 + 2.546530875 = 2.404780875$$

Nilai y (menggunakan $x^{(2)}=2.404780875$):
$$y^{(2)} = -0.05 y^{(1)} + 1.05\left[\frac{29-x^{(2)}}{10}\right]$$
$$y^{(2)} = -0.05(2.747325) + 1.05\left[\frac{29-2.404780875}{10}\right] = -0.13736625 + 1.05\left[\frac{26.595219125}{10}\right]$$
$$y^{(2)} = -0.13736625 + 1.05 \times 2.6595219125 = -0.13736625 + 2.792498008125 = 2.655131758125$$

Iterasi dilanjutkan terus hingga mencapai toleransi error yang diinginkan (misalnya, $e < 0.001$).

Untuk ringkasnya kita gunakan kode program berikut.

In [ ]:
# Inisialisasi
x0 = 0.0
y0 = 0.0
iter_count = 1 # Diubah dari 'iter' untuk menghindari konflik dengan fungsi bawaan

# Definisikan ulang fungsi GS di dalam scope ini agar jelas
f1_gs_calc = lambda x_curr, y_curr: (27.0 - y_curr) / 10.0
f2_gs_calc = lambda x_curr, y_curr: (29.0 - x_curr) / 10.0 # y_curr tidak digunakan di sini, x_curr adalah yang terpenting

# Masukan toleransi eror dan faktor relaksasi
e = float(input("Masukkan nilai toleransi error: "))
w = float(input("Masukkan nilai faktor relaksasi: "))

# SOR
print('\nIterasi\tx\ty\n')

condition = True
max_iter = 100 # Tambahkan batas maksimum iterasi untuk mencegah loop tak terbatas

while condition and iter_count <= max_iter:
    # Hitung x1 menggunakan y0 (nilai y dari iterasi sebelumnya)
    x1 = (1 - w) * x0 + w * f1_gs_calc(x0, y0) 
    # Hitung y1 menggunakan x1 (nilai x yang baru dihitung pada iterasi ini)
    y1 = (1 - w) * y0 + w * f2_gs_calc(x1, y0) # y0 di f2_gs_calc tidak dipakai, x1 yang dipakai
    
    print('%d\t%0.4f\t%0.4f\n' %(iter_count, x1, y1))
    
    # Cek konvergensi
    # Menggunakan norma maksimum dari perbedaan relatif atau absolut
    # Di sini kita gunakan perbedaan absolut seperti pada contoh sebelumnya
    e1 = abs(x0 - x1)
    e2 = abs(y0 - y1)
    
    iter_count += 1
    x0 = x1
    y0 = y1
    
    condition = (e1 > e or e2 > e) # Lanjutkan jika salah satu error masih di atas toleransi

if iter_count > max_iter and condition:
    print("\nMetode tidak konvergen dalam %d iterasi." % max_iter)
print('\nBeban kerja: Server A (x) = %0.3f, Server B (y) = %0.3f\n'% (x1, y1))

Masukkan nilai toleransi error: 0.0001
Masukkan nilai faktor relaksasi: 1.05

Iterasi	x	y

1	2.8350	2.7473

2	2.4048	2.6551

3	2.4350	2.6566

4	2.4343	2.6566

5	2.4343	2.6566


Beban kerja: Server A (x) = 2.434, Server B (y) = 2.657



## Pertanyaan

Lalu bagaimana jika kita mengubah nilai toleransi error dan faktor relaksasinya pada persamaan yang sama? Apakah hasil akan berbeda?

Gunakan kode di atas untuk menemukan jawabannya.

### Hasil pengamatan (untuk kasus Load Balancing)

Solusi analitik untuk sistem ini adalah:
$10x + y = 27$
$x + 10y = 29$
Dari (1), $y = 27 - 10x$. Substitusi ke (2):
$x + 10(27 - 10x) = 29$
$x + 270 - 100x = 29$
$-99x = -241$
$x = 241/99 \approx 2.434343...$
$y = 27 - 10(241/99) = (2673 - 2410)/99 = 263/99 \approx 2.656565...$

Beberapa percobaan dari kode program di atas untuk kasus Load Balancing dapat dirangkum sebagai berikut (hasil mungkin sedikit bervariasi tergantung implementasi error check yang presisi):

| e (Toleransi) | ω (Relaksasi) | → | x (Server A) | y (Server B) | Iterasi |
|---|---|---|---|---|---|
| 0.01 | 1.0 (Gauss-Seidel) |  | 2.43 | 2.66 | 6-7 |
| 0.01 | 1.05 |  | 2.43 | 2.66 | 5-6 |
| 0.01 | 1.1 |  | 2.43 | 2.66 | 6-7 |
| 0.01 | 0.9 |  | 2.44 | 2.66 | 8-9 |
| 0.001 | 1.05 |  | 2.434 | 2.657 | 7-8 |
| 0.001 | 1.07 (optimal teoritis untuk matriks ini sekitar 1.07-1.08) |  | 2.434 | 2.657 | 6-7 |
| 0.0001 | 1.05 |  | 2.4343 | 2.6566 | 9-10 |
| 0.0001 | 1.07 |  | 2.4343 | 2.6566 | 8-9 |

Dari semua data tersebut, bisa kita lihat bahwa nilai x dan y berada di kisaran $x \approx 2.434$ dan $y \approx 2.657$.

Jika melihat jumlah iterasi, nilai ω yang optimal (memberikan iterasi paling sedikit untuk akurasi tertentu) untuk sistem ini tampaknya berada sedikit di atas 1.0, misalnya sekitar 1.05 hingga 1.08. Nilai ω = 1.0 (Gauss-Seidel) konvergen, tetapi over-relaxation (ω > 1) dapat mempercepatnya. Under-relaxation (ω < 1) cenderung memperlambat konvergensi untuk sistem ini.

### Kesimpulan

Perubahan nilai toleransi error (e) dan faktor relaksasi (ω) pada metode SOR memengaruhi akurasi dan jumlah iterasinya, yaitu:
- Semakin kecil nilai **e (toleransi error)**, semakin akurat hasilnya, tetapi umumnya membutuhkan **lebih banyak iterasi**.
- Nilai **ω (faktor relaksasi)** sangat mempengaruhi kecepatan konvergensi:
  - **ω = 1** (Metode Gauss-Seidel) memberikan kecepatan konvergensi dasar.
  - **1 < ω < 2** (Over-relaxation) dapat mempercepat konvergensi secara signifikan jika ω dipilih mendekati nilai optimalnya. Untuk kasus ini, ω sekitar 1.05 - 1.08 tampaknya efektif.
  - **0 < ω < 1** (Under-relaxation) umumnya memperlambat konvergensi untuk sistem yang sudah konvergen dengan baik menggunakan Gauss-Seidel, tetapi bisa membantu untuk sistem yang sulit konvergen.
  - Memilih ω yang terlalu besar (mendekati 2 atau lebih) atau terlalu jauh dari optimal dapat memperlambat konvergensi atau bahkan menyebabkannya divergen.

Untuk kasus Load Balancing ini, solusi yang ditemukan adalah:
- **Beban kerja Server A (x) ≈ 2.434 unit**
- **Beban kerja Server B (y) ≈ 2.657 unit**

Nilai paling akurat bisa didapat jika:
- Hasil konvergen ke nilai stabil yang konsisten pada beberapa iterasi terakhir.
- Toleransi error (e) yang kecil digunakan.
- Faktor relaksasi (ω) dipilih secara optimal untuk meminimalkan jumlah iterasi sambil tetap memastikan konvergensi.